Tutorial for simulating microstructures (#127)
festim-dev/FESTIM-workshop: book/content/applications/microstructure.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.17.3
kernelspec:
  display_name: festim-workshop
  language: python
  name: python3
---

# Hydrogen diffusion along grain boundaries

This tutorial shows how to use FESTIM to simulate hydrogen diffusion in metal microstructures.

We'll show how to generate a microstructure using Voronoi cells, mesh it with GMSH, and solve a transport problem with FESTIM.

+++

## Geometry

### Generating the microstructure

First, we use `scipy` to make a [Voronoi diagram](https://en.wikipedia.org/wiki/Voronoi_diagram) mimicking a microstructure.

```{code-cell} ipython3
import numpy as np
from scipy.spatial import Voronoi, voronoi_plot_2d
import matplotlib.pyplot as plt
import gmsh

np.random.seed(1)

# N random seeds in [0,size]x[0,size]
size = 1.5
N_seeds = 150
points = np.random.rand(N_seeds, 2) * size

# centre everything on 0.5, 0.5
points -= size / 2
points += 0.5

vor = Voronoi(points)

voronoi_plot_2d(vor, show_vertices=False, show_points=True)
plt.show()
```

> Now that we have vertices for each voronoi cell, we can use `shapely` to turn them into `shapely.Polygon` objects for easy manipulation.

Reviewer: ```suggestion
Now that we have vertices for each Voronoi cell, we can use `shapely` to turn them into `shapely.Polygon` objects for easy manipulation.
```

> We then shrink the voronoi cells to make grain boundaries appear. Note that the grain boundary thickness is arbitrary here. 

Reviewer: ```suggestion
We then shrink the Voronoi cells to make grain boundaries appear. Note that the grain boundary thickness is arbitrary here. 
```

```{code-cell} ipython3
from shapely.geometry import Polygon
from shapely import difference, union_all
from shapely.plotting import plot_polygon, plot_points

gap = 0.01  # thickness of grain boundary

grains = []
for region_idx in vor.point_region:
    region = vor.regions[region_idx]
    if -1 in region or len(region) == 0:
        continue  # skip infinite regions
    poly = Polygon([vor.vertices[i] for i in region])
    poly = poly.intersection(
        Polygon([(0, 0), (1, 0), (1, 1), (0, 1)])
    )  # clip to domain
    if not poly.is_empty:
        grains.append(poly.buffer(-gap / 2))  # shrink each grain inward

print("Number of grains:", len(grains))

plt.figure(figsize=(8, 8))
# plot grains
for polygon in grains:
    plot_polygon(polygon=polygon, add_points=False)
    plot_points(polygon, markersize=2, color="tab:blue")
plt.show()
```

> To make a new `Polygon` for our grain boundaries by substracting all the grains from a square polygon.

Reviewer: ```suggestion
We can make a new `Polygon` for our grain boundaries by subtracting all the grains from a square polygon.
```

        point_coords.append(coords)

    # Check if both bounding points are on the left boundary (x ≈ eps)
    is_on_left = True
    for coords in point_coords:
        x_coord = coords[0]
        if abs(x_coord - eps) > 1e-9:
            is_on_left = False
            break

    # Check if both bounding points are on the right boundary (x ≈ 1-eps)
    is_on_right = True
    for coords in point_coords:
        x_coord = coords[0]
        if abs(x_coord - (1 - eps)) > 1e-9:
            is_on_right = False
            break

    if is_on_left:
        lines_on_left.append(tag)
    if is_on_right:
        lines_on_right.append(tag)

print("Lines on left boundary:", lines_on_left)
print("Lines on right boundary:", lines_on_right)

if lines_on_left:
    gmsh.model.addPhysicalGroup(1, lines_on_left, name="left_boundary")
if lines_on_right:
    gmsh.model.addPhysicalGroup(1, lines_on_right, name="right_boundary")

gmsh.model.mesh.generate(2)
# gmsh.fltk.run()
gmsh.write("voronoi_grains.msh")

gmsh.finalize()
```

## FESTIM model

We can now define our hydrogen transport problem.

$$
\frac{\partial c}{\partial t} = \nabla \cdot (D \nabla c)
$$

> where $D=D_\mathrm{GB}$ and $D=D_\mathrm{grain}$ in the grain boundary and in the grain, respectively.

Reviewer: ```suggestion
where $D=D_\mathrm{GB}$ and $D=D_\mathrm{grain}$ in the grain boundary and grain, respectively.
```

Boundary conditions:

- $ c = 1 $ on the left surface

> - $ c = 0 $ on the right surface

Author: ```suggestion
- $ c = 0 $ on the left surface
- $ c = 1 $ on the right surface
```

> - $ c = 0 $ on the right surface

Reviewer: I think the boundary conditions are flipped below

Interface condition:
- We assume continuity of concentration at the GB/grain interfaces.

> Simulation time: $t_f = 1.5$

Reviewer: ```suggestion
Simulation time: $t_f = 1.5$ seconds
```

Author: I don't want to use units here because i haven't used units for other things (like Diffusivity, c=1: 1 what?) and I don't want to because c = 1 H/m3 doesn't make any sense.
